### DISCUSSION TITLE
How to bind the value of a field in a query?

### DISCUSSION THREAD

**Discussion by brendonky18** [github/codeql#11354] [Q&A]
Looking at the [documentation](https://codeql.github.com/docs/ql-language-reference/types/#fields) on fields. It shows the following class with the divisor field:
```
class SmallInt extends int {
  SmallInt() { this = [1 .. 10] }
}

class DivisibleInt extends SmallInt {
  SmallInt divisor;   // declaration of the field `divisor`
  DivisibleInt() { this % divisor = 0 }

  SmallInt getADivisor() { result = divisor }
}
```

I wrote the following query, and it produced some unexpected results.
```
from DivisibleInt i
where i.getADivisor() = 2
select i, i.getADivisor()
```

I was expecting to get all of the even numbers, with just the divisor 2. Instead it gave me all of the even numbers and each of their possible divisors. 

![image](https://user-images.githubusercontent.com/20930961/203110665-9f4d487c-8a40-45df-87e2-8f7f9bca181f.png)

Can someone explain why this is happening, and how I could possibly fix my query/class to get the correct result?

**Comment by atorralba** [comment]
Note that `getADivisor()` gets _all_ the `SmallInt`s that give 0 as a result of the modulo operation with `this`. So what your query is currently doing is selecting tuples of a `DivisibleInt`s that has a divisor of 2, and all its divisors (one row for each).

So, what you actually want is binding the divisor to be the same one you are requiring to be equal to 2. Something like:

```ql
from DivisibleInt i, SmallInt divisor
where i.getADivisor() = divisor and divisor = 2
select i, divisor
```

**Reply by brendonky18** [reply]
So how would I go about doing this in a child class. Say I wanted to create a child class which only contains the tuples of the element, and the divisor of 2. Would there be a way to accomplish this without overriding the parent's predicate?

**Reply by atorralba** [reply]
To simply create a class that represents an `int` that is divisible by 2, you can do:

```ql
class DivisibleBy2Int extends DivisibleInt {
  DivisibleBy2Int() { divisor = 2 } // you're saying that this number must have 2 as one of its divisors
}
```

The `getADivisior` predicate as-is always gets all the divisors of the number, so you'd need to create a new predicate that does what you want (something like `hasDivisor(int i)` maybe).

Now, the thing is that you want to select a constant, so you can always do:

```ql
from DivisibleInt i
where i.getADivisor() = 2
select i, 2

// or

from DivisibleBy2Int i
select i, 2
```

(which is a little redundant)
